osquery/tables/system/windows/logon_sessions.cpp

```diff
@@ -0,0 +1,142 @@
+  adjust.QuadPart = JAN_1_1601_TO_JAN_1_1970;
+  date.QuadPart -= adjust.QuadPart;
+  return date.QuadPart / HUNDREDNANOSECONDS_TO_SECONDS;
+}
+
+LPWSTR GetLogonType(ULONG logonType) {
+  switch (logonType) {
+  case 2:
+    return L"Interactive";
+  case 3:
+    return L"Network";
+  case 4:
+    return L"Batch";
+  case 5:
+    return L"Service";
+  case 6:
+    return L"Proxy";
+  case 7:
+    return L"Unlock";
+  case 8:
+    return L"NetworkCleartext";
+  case 9:
+    return L"NewCredentials";
+  case 10:
+    return L"RemoteInteractive";
+  case 11:
+    return L"CachedInteractive";
+  case 12:
+    return L"CachedRemoteInteractive";
+  case 13:
+    return L"CachedUnlock";
+  default:
+    return L"None";
+  }
+}
+
+QueryData QueryLogonSessions(QueryContext& context) {
+  QueryData results;
+  bool loadedManually = false;
+
+  HMODULE module = GetModuleHandle(TEXT("Secur32.dll"));
+
+  if (!module) {
+    module = LoadLibrary(TEXT("Secur32.dll"));
+    loadedManually = true;
+  }
+
+  NTSTATUS(__stdcall * LsaEnumerateLogonSessions)
+  (PULONG LogonSessionCount, PLUID * LogonSessionList);
+  LsaEnumerateLogonSessions =
+      reinterpret_cast<decltype(LsaEnumerateLogonSessions)>(
+          GetProcAddress(module, "LsaEnumerateLogonSessions"));
+
+  NTSTATUS(__stdcall * LsaGetLogonSessionData)
+  (PLUID LogonId, PSECURITY_LOGON_SESSION_DATA * ppLogonSessionData);
+  LsaGetLogonSessionData = reinterpret_cast<decltype(LsaGetLogonSessionData)>(
+      GetProcAddress(module, "LsaGetLogonSessionData"));
+
+  NTSTATUS(__stdcall * LsaFreeReturnBuffer)(PVOID Buffer);
+  LsaFreeReturnBuffer = reinterpret_cast<decltype(LsaFreeReturnBuffer)>(
+      GetProcAddress(module, "LsaFreeReturnBuffer"));
+
+  if (LsaEnumerateLogonSessions && LsaGetLogonSessionData &&
+      LsaFreeReturnBuffer) {
+    PLUID sessions = NULL;
+    ULONG sessionCount = 0;
+    NTSTATUS status = LsaEnumerateLogonSessions(&sessionCount, &sessions);
+
+    if (status == 0) {
+      for (size_t i = 0; i < sessionCount; i++) {
+        PSECURITY_LOGON_SESSION_DATA sessionData = NULL;
+        NTSTATUS status = LsaGetLogonSessionData(&sessions[i], &sessionData);
+
+        if (status == 0) {
+          Row r;
+          r["logon_id"] = INTEGER(sessionData->LogonId.LowPart);
+          r["user"] = wstringToString(sessionData->UserName.Buffer);
+          r["logon_domain"] = wstringToString(sessionData->LogonDomain.Buffer);
+          r["authentication_package"] =
+              wstringToString(sessionData->AuthenticationPackage.Buffer);
+          r["logon_type"] =
+              wstringToString(GetLogonType(sessionData->LogonType));
+          r["session_id"] = INTEGER(sessionData->Session);
+          LPTSTR sid;
+          if (ConvertSidToStringSid(sessionData->Sid, &sid)) {
+            r["logon_sid"] = sid;
+          }
+          if (sid) {
+            LocalFree(sid);
+          }
+          r["logon_time"] = BIGINT(filetimeToUnixtime(sessionData->LogoffTime));
```

Comment:
You might want to use sessionData->LogonTime

Reply:
Good catch! I'll make that change now 👍

Reply:
Any reason LogoffTime isn't also included in the table? I'm specifically needing to be able to look at session duration / time of day, and was disappointed to find it missing!

Reply:
I personally can't recall a reason why it wasn't included, but it should be simple enough to add it for a new release.